### PR TITLE
feat(navbar): Add themepicker component with lazy loaded themes

### PR DESCRIPTION
**Pull request by riavalon**
Picking up the theme picker feature from where it was left off.

### CODE REVIEW

File: src/app/shared/style-manager/style-manager.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    let className = `style-manager-${key}`;
+    let linkEl = document.head.querySelector(`link[rel="stylesheet"].${className}`);
+    if (!linkEl && create) {
+      linkEl = document.createElement('link');
+      linkEl.setAttribute('rel', 'stylesheet');
+      linkEl.classList.add(className);
+      document.head.appendChild(linkEl);
+    }
+    return linkEl as HTMLLinkElement;
+  }
```
Comment (jelbourn) [inline src/app/shared/style-manager/style-manager.ts]: Rather than having one function perform two behaviors with an option, I generally prefer breaking it up into separate functions. Also, most of these can be module-level functions since they are pure functions, e.g.
```ts
class StyleManager {
  setStyle(key: string, href: string) {
    getLinkElementForKey(key).setAttribute('href', href);
  }

  removeStyle(key: string) {
    const existingLinkElement = getExistingLinkElementByKey(key);
    if (existingLinkElement) {
      document.head.removeChild(existingLinkElement);
    }
  }
}

function getLinkElementForKey(key: string) {
  return getExistingLinkElementByKey(key) || createLinkElementWithKey(key);
}

function getExistingLinkElementByKey(key: string) {
  return document.head.querySelector(`link[rel="stylesheet"].${getClassNameForKey(key)}`);
}

function createLinkElementWithKey(key: string) {
  const linkEl = document.createElement('link');
  linkEl.setAttribute('rel', 'stylesheet');
  linkEl.classList.add(getClassNameForKey(key));
  document.head.appendChild(linkEl);

  return linkEl;
}

function getClassNameForKey(key: string) {
  return `style-manager-${key}`;
}
```

File: src/app/shared/theme-chooser/theme-chooser.scss [hunk 2]
```diff
@@ -0,0 +1,52 @@
+$theme-chooser-menu-padding: 8px;
+$theme-chooser-grid-cell-size: 48px;
+$theme-chooser-grid-cells-per-row: 2;
+$theme-chooser-swatch-size: 36px;
+$theme-chooser-accent-stripe-size: 6px;
+
+
+.theme-chooser-menu {
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.scss]: Prefix css classes with `docs-`

File: src/app/shared/theme-chooser/theme-chooser.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import {Component, ViewEncapsulation, ChangeDetectionStrategy} from '@angular/core';
+import {StyleManager} from '../style-manager/style-manager';
+
+@Component({
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.ts]: We should make the theme chooser `aria-hidden="true"` and ensure that the button is not in the tab order since a screen-reader user won't get any benefit from changing the app colors.

File: src/app/shared/theme-chooser/theme-chooser.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  styleUrls: ['theme-chooser.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+})
+export class ThemeChooser {
+  themes = [
+    {
+      primary: '#673AB7',
+      accent: '#FFC107',
+      href: 'assets/deeppurple-amber.css'
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.ts]: We should omit `assets/` on each entry and add it in the function call.

File: angular-cli.json [hunk 5]
```diff
@@ -19,7 +19,11 @@
       "mobile": false,
       "styles": [
         "main.scss",
-        "highlightjs/solarized-light.css"
+        "highlightjs/solarized-light.css",
+        {"input": "assets/pink-bluegrey.css", "lazy": true},
+        {"input": "assets/deeppurple-amber.css", "lazy": true},
+        {"input": "assets/indigo-pink.css", "lazy": true},
+        {"input": "assets/purple-green.css", "lazy": true}
```
Comment (jelbourn) [inline angular-cli.json]: Can these be referenced directly from `node_modules` rather than copying them to assets?

File: src/app/shared/theme-chooser/theme-chooser.scss [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  .theme-chooser-primary {
+    width: 100%;
+    height: 100%;
+  }
+
+  .theme-chooser-accent {
+    position: absolute;
+    bottom: $theme-chooser-accent-stripe-size;
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.scss]: How do you feel about the accent stripe? I'm not particularly fond of it.

File: src/app/app-module.ts [hunk 7]
```diff
@@ -32,7 +35,14 @@ import {ComponentPageHeader} from './pages/component-page-header/component-page-
     GuideList,
     GuideViewer,
     Homepage,
-    Footer
+    Footer,
+  ],
+  schemas: [
+    CUSTOM_ELEMENTS_SCHEMA,
```
Comment (jelbourn) [inline src/app/app-module.ts]: `CUSTOM_ELEMENTS_SCHEMA` ideally shouldn't be here; it why was this needed?

File: src/app/app-module.ts [hunk 8]
```diff
@@ -44,10 +54,12 @@ import {ComponentPageHeader} from './pages/component-page-header/component-page-
     MaterialModule,
     MdNativeDateModule,
     routing,
+    InlineSVGModule,
```
Comment (jelbourn) [inline src/app/app-module.ts]: `InlineSvgModule` (camelCase even for acronyms)

File: src/app/material-docs-app.ts [hunk 9]
```diff
@@ -1,6 +1,8 @@
 import {Component, ViewEncapsulation} from '@angular/core';
 import {Router, NavigationStart} from '@angular/router';
 
+// import {ThemeStorage} from './shared/theme-chooser/theme-storage/theme-storage';
```
Comment (jelbourn) [inline src/app/material-docs-app.ts]: Commented line

File: src/app/material-docs-app.ts [hunk 10]
```diff
@@ -15,9 +17,19 @@ export class MaterialDocsApp {
...
     });
+
+    // _themeStorage
+    //   .onThemeUpdate
+    //   .subscribe((theme) => this._setIsDarkTheme(theme));
   }
+
+  // public _setIsDarkTheme(theme) {
+  //   this.isDarkTheme = theme ? theme.isDark : this.isDarkTheme;
+  // }
```
Comment (jelbourn) [inline src/app/material-docs-app.ts]: Commented code

File: src/app/pages/component-category-list/component-category-list.scss [hunk 11]
```diff
@@ -22,12 +22,9 @@
   }
 }
 
-.docs-component-category-list-card-image {
-  width: 100%;
-  height: 160px;
-  background-size: contain;
-  background-repeat: no-repeat;
-  background-position: center;
+:host /deep/ .docs-component-category-list-card-image svg {
```
Comment (jelbourn) [inline src/app/pages/component-category-list/component-category-list.scss]: I'd like to avoid `/deep/` completely, instead preferring to add a global style (or, barring that, turning off view encapsulation for the component)

File: src/app/pages/component-category-list/component-category-list.scss [hunk 12]
```diff
@@ -22,12 +22,9 @@
...
+  width: 100% !important;
+  height: 100% !important;
```
Comment (jelbourn) [inline src/app/pages/component-category-list/component-category-list.scss]: Why is `!important` needed here? Can this be done differently?

File: src/app/pages/component-sidenav/_component-sidenav-theme.scss [hunk 13]
```diff
@@ -1,4 +1,4 @@
-@import '~@angular/material/theming';
+@import '../../../../node_modules/@angular/material/theming';
```
Comment (jelbourn) [inline src/app/pages/component-sidenav/_component-sidenav-theme.scss]: You should be able to change all of these imports back; there was a short-lived bug in webpack's scss loader that made the `~` syntax not work that should be fixed with a fresh `npm install`

File: src/app/pages/homepage/homepage.html [hunk 14]
```diff
@@ -13,7 +13,8 @@
 <div class="docs-homepage-promo">
   <div class="docs-homepage-row">
     <div class="docs-homepage-promo-img">
-      <img src="../assets/img/homepage/sprintzerotoapp.svg" alt="Sprint from Zero to App" />
+      <docs-svg-viewer svgHref="../assets/img/homepage/sprintzerotoapp.svg"></docs-svg-viewer>
+      <!-- <div inlineSVG="../assets/img/homepage/sprintzerotoapp.svg" alt="Sprint from Zero to App"></div> -->
```
Comment (jelbourn) [inline src/app/pages/homepage/homepage.html]: Commented code (here and below)

File: src/app/shared/navbar/navbar.scss [hunk 15]
```diff
@@ -10,14 +11,18 @@
   }
 }
 
-.docs-angular-logo {
+.docs-angular-logo /deep/ svg {
```
Comment (jelbourn) [inline src/app/shared/navbar/navbar.scss]: Same comment about `/deep/`

File: src/app/shared/shared-module.ts [hunk 16]
```diff
@@ -1,14 +1,20 @@
...
+import {ExampleViewer} from './example-viewer/example-viewer';
+import {DocViewer} from './doc-viewer/doc-viewer';
+import {DocumentationItems} from './documentation-items/documentation-items';
 import {PlunkerButton} from './plunker';
 import {GuideItems} from './guide-items/guide-items';
+import {ThemeStorage} from './theme-chooser/theme-storage/theme-storage';
+import {ThemeChooser} from './theme-chooser/theme-chooser';
+import {NavBar} from './navbar/navbar';
+import {SVGBuilder} from './svg-viewer/svg-builder';
+import {SVGViewer} from './svg-viewer/svg-viewer';
```
Comment (jelbourn) [inline src/app/shared/shared-module.ts]: `SvgBuilder`, `SvgViewer`

File: src/app/shared/svg-viewer/svg-builder.ts [hunk 17]
```diff
@@ -0,0 +1,73 @@
+import {Injectable} from '@angular/core';
+
+import {IDocsSiteTheme} from '../theme-chooser/theme-storage/theme-storage';
+
+const Color = require('color');
+
+
+export interface IThemeColors {
```
Comment (jelbourn) [inline src/app/shared/svg-viewer/svg-builder.ts]: We prefer not to prefix interfaces with `I` (here and elsewhere)

File: src/app/shared/svg-viewer/svg-builder.ts [hunk 18]
```diff
@@ -0,0 +1,73 @@
+import {Injectable} from '@angular/core';
+
+import {IDocsSiteTheme} from '../theme-chooser/theme-storage/theme-storage';
+
+const Color = require('color');
```
Comment (jelbourn) [inline src/app/shared/svg-viewer/svg-builder.ts]: What's this `require` doing here?

File: src/app/shared/svg-viewer/svg-viewer.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    public el: ElementRef
+  ) {
+    this.currTheme = this.themeStorage.getStoredTheme();
+    this._subscription = this.themeStorage.onThemeUpdate
+      .subscribe(theme => this.swapTheme(theme));
+  }
+
+  public ngAfterViewInit() {
+    if (this.currTheme) {
+      setTimeout(this.swapTheme.bind(this, this.currTheme));
```
Comment (jelbourn) [inline src/app/shared/svg-viewer/svg-viewer.ts]: Prefer arrow functions to using `bind`

File: src/app/shared/theme-chooser/theme-chooser.html [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+</button>
+
+<md-menu class="docs-theme-chooser-menu" #themeMenu="mdMenu" x-position="before">
+  <md-grid-list cols="2">
+    <md-grid-tile *ngFor="let theme of themes">
+      <div md-menu-item (click)="installTheme(theme.href)">
+        <div class="docs-theme-chooser-swatch">
+          <md-icon class="docs-theme-chosen-icon" *ngIf="currentTheme === theme">check_circle</md-icon>
+          <div class="docs-theme-chooser-primary" [style.background]="theme.primary"></div>
+          <!--<div class="docs-theme-chooser-accent" [style.background]="theme.accent"></div>-->
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.html]: Commented code

File: src/app/shared/theme-chooser/theme-chooser.html [hunk 21]
```diff
@@ -0,0 +1,17 @@
+<button md-icon-button [md-menu-trigger-for]="themeMenu">
+  <md-icon>format_color_fill</md-icon>
+</button>
+
+<md-menu class="docs-theme-chooser-menu" #themeMenu="mdMenu" x-position="before">
```
Comment (jelbourn) [inline src/app/shared/theme-chooser/theme-chooser.html]: Add a TODO here like
```html
<!-- TODO: replace use of `md-menu` here with a custom overlay -->
```

File: src/app/pages/component-sidenav/_component-sidenav-theme.scss (hunk 13 above)
Comment (riavalon) [inline src/app/pages/component-sidenav/_component-sidenav-theme.scss]: This doesn't quite appear to work still. After removing node_modules, running `npm cache clean` and doing a fresh install, I'm still seeing errors trying to resolve the `~angular/material/theming` paths.

Comment (riavalon) [inline src/app/pages/component-sidenav/_component-sidenav-theme.scss]: @jelbourn  As an update on this issue, compiling the individual `scss` files for each theme doesn't have access to webpack, but instead uses `node-sass`, so the `~` operator will not work here unfortunately.

**Comment by masonlr** [comment]
@jelbourn I'm looking at `theme-picker.ts` and can't see how the default theme (`isDefault: true` property in `themes`) is honoured when you first load the page.

`installTheme()` is only called if there is a theme in local storage or a theme is supplied by query parameters. If it's the first time you're visiting the page, neither of these may be present.
